File: app/src/main/java/com/teamunemployment/lolanalytics/Login/LoginModel.java
package com.teamunemployment.lolanalytics.Login;

import com.teamunemployment.lolanalytics.Data.RESTApiExecutor;
import com.teamunemployment.lolanalytics.Data.RealmExecutor;
import com.teamunemployment.lolanalytics.Data.model.LongWrapper;

import javax.inject.Inject;


import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;


/**
 * @author Josiah Kendall
 */

public class LoginModel {

    private RESTApiExecutor restApiExecutor;
    private RealmExecutor realmExecutor;

    @Inject
    public LoginModel(RESTApiExecutor restApiExecutor, RealmExecutor realmExecutor) {
        this.restApiExecutor = restApiExecutor;
        this.realmExecutor = realmExecutor;
    }


    public void requestLoginAttempt(String summonerName, String region, final LoginContract.BasePresenter presenter) {
        Observable<LongWrapper> longWrapperObservable = restApiExecutor.GetSummonerId(summonerName, region);
        longWrapperObservable.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<LongWrapper>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(LongWrapper value) {
                        presenter.setLoginResult(value.data);
                    }

                    @Override
                    public void onError(Throwable e) {
                        presenter.handleError(e);
                    }

                    @Override
                    public void onComplete() {

                    }
                });

    }

    public void saveUserDetails(String username, long loginResult, String region) {

    }
}

File: app/src/main/java/com/teamunemployment/lolanalytics/login/LoginModel.kt
package com.teamunemployment.lolanalytics.login

import co.metalab.asyncawait.async
import com.teamunemployment.lolanalytics.Data.model.LongWrapper
import com.teamunemployment.lolanalytics.io.RealmExecutor

import javax.inject.Inject


/**
 * @author Josiah Kendall
 */

class LoginModel @Inject
constructor(private val loginRestService: LoginRestService, private val realmExecutor: RealmExecutor) {

    /**
     * Login a user.
     * @param summonerName The name of the summoner to search for
     */
    fun login(summonerName: String, region: String, presenter: LoginContract.Presenter){
        async {
            val result : LongWrapper = await { loginRestService.doesUserExist(summonerName, region) }

            if (result.data != 200.toLong()) {
                presenter.handleLoginResult(result.data)
            } else {
                // save user to database
                val syncResult = await { loginRestService.beginSync(summonerName, region) }
                presenter.handleLoginResult(syncResult.data)
            }
        }
    }
}
